progress towards implementing proper ARP in reachability analysis

arifogel opened the pull request:
- generalized IpSpace into an interface
- removed all mention of flow sinks
- commented out broken tests
- implemented new ArpTrueEdge analysis

arifogel commented:
Review status: 0 of 34 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/batfish/batfish/1047#-L7RpMAGoOnCZNYcDx8H:-L7RpMAHIzn8EBbZo7eE:bq5zy57) (no related file):*
DRAFT ONLY. NOT TO BE MERGED.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 28 of 34 files at r1, 5 of 5 files at r2.
Review status: 33 of 34 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/GenericRib.java, line 27 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VBAOKBPBPD0MpjybD:-L7VBAOM8cPEhVp2unPF:b-fy9kqv) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/GenericRib.java#L27)):*
> ```Java
> 
>   /** Get all the IPs for which there is a matching route */
>   IpWildcardSetIpSpace getRoutableIps();
> ```

Which client of `GenericRib` depends on this being `IpWildcardSetIpSpace` rather than `IpSpace` and why?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpAddressAcl.java, line 7 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VB_o70VFG8vAPp1t0:-L7VB_o70VFG8vAPp1t1:b-5mlgd3) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpAddressAcl.java#L7)):*
> ```Java
> import javax.annotation.Nonnull;
> 
> public class IpAddressAcl implements IpSpace {
> ```

add comment: permit means in the space, deny means not in the space

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpAddressAclLine.java, line 16 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VBqnpBrFleSPr6sPd:-L7VBqnpBrFleSPr6sPe:b-ttmk9z) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpAddressAclLine.java#L16)):*
> ```Java
>     private IpSpace _ipSpace;
> 
>     private boolean _negate;
> ```

what is this for?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpWildcardSetIpSpace.java, line 10 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VCyvLDeEVKYmGQ1US:-L7VCyvLDeEVKYmGQ1UT:bn6ozsb) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpWildcardSetIpSpace.java#L10)):*
> ```Java
> 
> /** Represents a space of IPv4 addresses. */
> public final class IpWildcardSetIpSpace implements IpSpace, Serializable {
> ```

comment: uses a whitelist and a blacklist of ip wildcards. Blacklist takes priority, so if an Ip is matched by both lists, it is not in the space.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/FibRowMatchExpr.java, line 3 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VELrb58Zsr8Uv72dk:-L7VELrb58Zsr8Uv72dl:b-bcl2k) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish/src/main/java/org/batfish/z3/expr/FibRowMatchExpr.java#L3)):*
> ```Java
> package org.batfish.z3.expr;
> 
> public class FibRowMatchExpr {
> ```

can we remove this class now?

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 719 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VEwhR6jWwuedAc4qz:-L7VEwhR6jWwuedAc4r-:b-8to3nb) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L719)):*
> ```Java
> 
>   @Override
>   public void visitPreInInterface(PreInInterface.State preInInterface) {
> ```

did the code just move, or did it change too?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 21 of 33 files reviewed at latest revision, 7 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpAddressAcl.java, line 7 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VB_o70VFG8vAPp1t0:-L7Wxvyp6l7zifWbNvMd:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpAddressAcl.java#L7)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

add comment: permit means in the space, deny means not in the space
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpAddressAclLine.java, line 16 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VBqnpBrFleSPr6sPd:-L7WyOq90i4hm63DEeoo:bchcon) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpAddressAclLine.java#L16)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

what is this for?
</blockquote></details>

Renamed and added comment

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpWildcardSetIpSpace.java, line 10 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VCyvLDeEVKYmGQ1US:-L7WyhAmCyESLmovGwhS:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/IpWildcardSetIpSpace.java#L10)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

comment: uses a whitelist and a blacklist of ip wildcards. Blacklist takes priority, so if an Ip is matched by both lists, it is not in the space.
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/FibRowMatchExpr.java, line 3 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VELrb58Zsr8Uv72dk:-L7WyhyPES_QbZOAn5n3:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish/src/main/java/org/batfish/z3/expr/FibRowMatchExpr.java#L3)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

can we remove this class now?
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/GenericRib.java, line 27 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VBAOKBPBPD0MpjybD:-L7Wxa06Bt1eTmmKGSk7:btyn42f) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/GenericRib.java#L27)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Which client of `GenericRib` depends on this being `IpWildcardSetIpSpace` rather than `IpSpace` and why?
</blockquote></details>

Good catch. Changed to `IpSpace`.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 21 of 33 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 719 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VEwhR6jWwuedAc4qz:-L7_aH55B_1IKZLVrAAn:b8w623e) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L719)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

did the code just move, or did it change too?
</blockquote></details>

Don't bother reviewing this file for now. I am probably going to recopy the contents from master before making necessary changes.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 13 of 13 files at r3.
Review status: 32 of 33 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/DataPlaneArpAnalysis.java, line 121 at r3](https://reviewable.io:443/reviews/batfish/batfish/1047#-L7XK8HxpuqnABqiIIBT:-L7XK8HxpuqnABqiIIBU:bbuq9n6) ([raw file](https://github.com/batfish/batfish/blob/398e7d5f09af850a9fa1adb0d9814b9e4e2a3a7c/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/DataPlaneArpAnalysis.java#L121)):*
> ```Java
>       Map<String, IpSpace> routableIpsByVrf,
>       SortedMap<String, GenericRib<AbstractRoute>> ribsByVrf,
>       Map<String, IpSpace> ipsRoutedOutInterfaces) {
> ```

what's the difference between this and `routableIpsByVrf`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 32 of 33 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/DataPlaneArpAnalysis.java, line 121 at r3](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7XK8HxpuqnABqiIIBT:-L7_rVgFCB9cw-lcLpdF:brlxi7w) ([raw file](https://github.com/batfish/batfish/blob/398e7d5f09af850a9fa1adb0d9814b9e4e2a3a7c/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/DataPlaneArpAnalysis.java#L121)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

what's the difference between this and `routableIpsByVrf`?
</blockquote></details>

`routableIpsByVrf` in this function maps the name of a VRF to the set of all IPs it can route.
The return value of this function maps the name of an interface (irrespective of what VRF it is in, since that can be determined by the interface) to the set of IPs routed out this interface in particular. It should be noted that the union of all such IPs for interfaces in a given VRF in the latter is not necessarily equal to `routableIpsByVrf` for that VRF, since there may also be null routes (no interface).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 32 of 33 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/DataPlaneArpAnalysis.java, line 121 at r3](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7XK8HxpuqnABqiIIBT:-L7_sGVC7DW1u_GiWncz:bgd94xx) ([raw file](https://github.com/batfish/batfish/blob/398e7d5f09af850a9fa1adb0d9814b9e4e2a3a7c/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/DataPlaneArpAnalysis.java#L121)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

`routableIpsByVrf` in this function maps the name of a VRF to the set of all IPs it can route.
The return value of this function maps the name of an interface (irrespective of what VRF it is in, since that can be determined by the interface) to the set of IPs routed out this interface in particular. It should be noted that the union of all such IPs for interfaces in a given VRF in the latter is not necessarily equal to `routableIpsByVrf` for that VRF, since there may also be null routes (no interface).
</blockquote></details>

Sorry, replace `return value of this function` -> ``ipsRoutedOutInterfaces` in above comment.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 29 of 33 files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 719 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7VEwhR6jWwuedAc4qz:-L7_xGhU4WNDQSyUB9DO:bvq04cc) ([raw file](https://github.com/batfish/batfish/blob/a8813a1b467b5ebfea09240bacf4fdb45250ef5b/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L719)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Don't bother reviewing this file for now. I am probably going to recopy the contents from master before making necessary changes.
</blockquote></details>

(Reverted to master version - suggest to just mark as viewed for now).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 7 of 8 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/DataPlaneArpAnalysis.java, line 121 at r3](https://reviewable.io:443/reviews/batfish/batfish/1047#-L7XK8HxpuqnABqiIIBT:-L7aMnWlrgGBYLsdqKe5:b-k4vwi1) ([raw file](https://github.com/batfish/batfish/blob/398e7d5f09af850a9fa1adb0d9814b9e4e2a3a7c/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/DataPlaneArpAnalysis.java#L121)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Sorry, replace `return value of this function` -> ``ipsRoutedOutInterfaces` in above comment.
</blockquote></details>

makes sense. please add that as a comment somewhere appropriate (and if not here, then refer to it here)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 20 of 47 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/DataPlaneArpAnalysis.java, line 121 at r3](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7XK8HxpuqnABqiIIBT:-L7b4u-F48gx4npwBGBU:b-ydx8hu) ([raw file](https://github.com/batfish/batfish/blob/398e7d5f09af850a9fa1adb0d9814b9e4e2a3a7c/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/DataPlaneArpAnalysis.java#L121)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

makes sense. please add that as a comment somewhere appropriate (and if not here, then refer to it here)
</blockquote></details>

Now I'm not sure the comment is quite correct regarding null interfaces. I think it will become clear as I fix the tests.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 27 of 27 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/ArpAnalysis.java, line 25 at r5](https://reviewable.io:443/reviews/batfish/batfish/1047#-L7erLz4DKVTBxOo3eqI:-L7erLz5zRC-HbLmDIK2:bh7dcd3) ([raw file](https://github.com/batfish/batfish/blob/1a569e10fea2025f44ee8394bbc40b4616961768/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/ArpAnalysis.java#L25)):*
> ```Java
> 
>   /** Mapping: hostname -> vrfName -> nullableIps */
>   Map<String, Map<String, IpSpace>> getNullableIps();
> ```

What is a nullable IP?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/visitors/GenericIpSpaceVisitor.java, line 10 at r5](https://reviewable.io:443/reviews/batfish/batfish/1047#-L7ershTz-BkZooUzgtA:-L7ershTz-BkZooUzgtB:b4uynhv) ([raw file](https://github.com/batfish/batfish/blob/1a569e10fea2025f44ee8394bbc40b4616961768/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/visitors/GenericIpSpaceVisitor.java#L10)):*
> ```Java
>   R castToGenericIpSpaceVisitorReturnType(Object o);
> 
>   R visitIpAddressAcl(IpAddressAcl ipAddressAcl);
> ```

Should we rename `IpAddressAcl` to `XIpSpace`? Perhaps `AclIpSpace`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 41 of 47 files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/ArpAnalysis.java, line 25 at r5](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7erLz4DKVTBxOo3eqI:-L7kWknTAcjLR3Ct4qKA:b-dw1mhg) ([raw file](https://github.com/batfish/batfish/blob/1a569e10fea2025f44ee8394bbc40b4616961768/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/ArpAnalysis.java#L25)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

What is a nullable IP?
</blockquote></details>

I added an explanation.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/visitors/GenericIpSpaceVisitor.java, line 10 at r5](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7ershTz-BkZooUzgtA:-L7kXvlj2s1nibiFp5lF:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/1a569e10fea2025f44ee8394bbc40b4616961768/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/visitors/GenericIpSpaceVisitor.java#L10)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Should we rename `IpAddressAcl` to `XIpSpace`? Perhaps `AclIpSpace`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 29 of 54 files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7la3XAEbV0igjQgcpy:-L7la3XAEbV0igjQgcpz:b-y2eecw) (no related file):*
TODO: finish SynthesizerInputImplTest

---

*[a discussion](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7la89l82pBvjE1oXuI:-L7la89l82pBvjE1oXuJ:bjdfoq) (no related file):*
TODO: write ArpAnalysis tests

---

*[a discussion](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7laBzHEu17HUA13t_k:-L7laBzHEu17HUA13t_l:b-z4jm5x) (no related file):*
TODO: fix compression tests

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 27 of 27 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/IBatfish.java, line 187 at r1](https://reviewable.io:443/reviews/batfish/batfish/1047#-L7zbtEVEWyCBcb6L6Vr:-L7zbtEVEWyCBcb6L6Vs:b-cyugl2) ([raw file](https://github.com/batfish/batfish/blob/6af8ebe1ee917fe827bcd42b31897fbadaa40219/projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/IBatfish.java#L187)):*
> ```Java
>       Set<String> transitNodes,
>       Set<String> notTransitNodes,
>       boolean useCompression,
> ```

why isn't `useCompression` part of `ReachabilitySettings`?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/ArpAnalysis.java, line 25 at r5](https://reviewable.io:443/reviews/batfish/batfish/1047#-L7erLz4DKVTBxOo3eqI:-L7zcIpSPIx8zro3u6E9:b-yma9nf) ([raw file](https://github.com/batfish/batfish/blob/1a569e10fea2025f44ee8394bbc40b4616961768/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/ArpAnalysis.java#L25)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

I added an explanation.
</blockquote></details>

I think this deserves a better name. Perhaps "unroutableIps" or "discardedIps" or "nullRoutedIps"?

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 1616 at r5](https://reviewable.io:443/reviews/batfish/batfish/1047#-L7zqMn1dnnxWaEyhLZG:-L7zqMn1dnnxWaEyhLZH:b-pp5nw5) ([raw file](https://github.com/batfish/batfish/blob/1a569e10fea2025f44ee8394bbc40b4616961768/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L1616)):*
> ```Java
>                 ImmutableSet.of(new PostInVrf(NODE1, VRF1), new PreOut(NODE1)),
>                 new PreOutEdge(NODE1, INTERFACE1, NODE2, INTERFACE2))));
>     assertThat(
> ```

what's replacing all of these tests?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 1 of 27 files at r6, 6 of 16 files at r8.
Review status: 50 of 60 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 1 of 27 files at r6, 10 of 16 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: 59 of 61 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[projects/checkstyle.xml, line 43 at r9](https://reviewable.io:443/reviews/batfish/batfish/1047#-L83O4xWvioyHSLOapnW:-L83O4xWvioyHSLOapnX:b-9fy2nl) ([raw file](https://github.com/batfish/batfish/blob/2e02a47764654cbe24fdf14346f0ce73ac7c967a/projects/checkstyle.xml#L43)):*
> ```XML
>             <property name="allowNonPrintableEscapes" value="true"/>
>         </module>
>         <!-- Autoformat of nested lambdas made this unworkable
> ```

could be an indication that you should refactor to avoid so many nested lambdas

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 1 of 2 files at r9.
Review status: 60 of 61 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 1 of 2 files at r9.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 60 of 61 files reviewed at latest revision, 7 unresolved discussions.

---

*[projects/checkstyle.xml, line 43 at r9](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L83O4xWvioyHSLOapnW:-L84GwNL0b0qwqykMffD:b-gdrbig) ([raw file](https://github.com/batfish/batfish/blob/2e02a47764654cbe24fdf14346f0ce73ac7c967a/projects/checkstyle.xml#L43)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

could be an indication that you should refactor to avoid so many nested lambdas
</blockquote></details>

Tomato potato

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 52 of 61 files reviewed at latest revision, 7 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/ArpAnalysis.java, line 25 at r5](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7erLz4DKVTBxOo3eqI:-L84IZ7I2k_44gSorhr5:b-6f7ux8) ([raw file](https://github.com/batfish/batfish/blob/1a569e10fea2025f44ee8394bbc40b4616961768/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/ArpAnalysis.java#L25)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

I think this deserves a better name. Perhaps "unroutableIps" or "discardedIps" or "nullRoutedIps"?
</blockquote></details>

Changed everywhere to `nullRoutedIps`

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 51 of 63 files reviewed at latest revision, 7 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/IBatfish.java, line 187 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7zbtEVEWyCBcb6L6Vr:-L84McWeEoQOPH6sbl5U:b-3grp88) ([raw file](https://github.com/batfish/batfish/blob/6af8ebe1ee917fe827bcd42b31897fbadaa40219/projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/IBatfish.java#L187)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

why isn't `useCompression` part of `ReachabilitySettings`?
</blockquote></details>

Not sure how I arrived at this confusing result. I moved `actions` and `useCompression` into `ReachabilitySettings`

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 51 of 63 files reviewed at latest revision, 7 unresolved discussions.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 1616 at r5](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L7zqMn1dnnxWaEyhLZG:-L84QOLKC79ts_WEWuow:b-cd3sf4) ([raw file](https://github.com/batfish/batfish/blob/1a569e10fea2025f44ee8394bbc40b4616961768/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L1616)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

what's replacing all of these tests?
</blockquote></details>

I had stopped for the day and forgot to add TODO. These tests should cover all forms of redunndancy.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 27 of 30 files at r10.
Review status: 71 of 72 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java, line 31 at r10](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88LEwSCiEoMRrT9CZM:-L88LEwSCiEoMRrT9CZN:b-aflr8e) ([raw file](https://github.com/batfish/batfish/blob/57dcf3603e4ba241a90a09658320ca493eb8bea7/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java#L31)):*
> ```Java
>     private boolean _active;
> 
>     private Set<InterfaceAddress> _allAddresses;
> ```

what is this?

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestArpAnalysis.java, line 7 at r6](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88LgPtCO-b97GAACPg:-L88LgPtCO-b97GAACPh:bx4jp42) ([raw file](https://github.com/batfish/batfish/blob/3e00658c6667934835c8270e6f5979b67f2d0d2d/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestArpAnalysis.java#L7)):*
> ```Java
> import java.util.Set;
> 
> public class TestArpAnalysis implements ArpAnalysis {
> ```

I expected to see tests here. Maybe rename it `MockArpAnalysis`?

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestIpSpace.java, line 6 at r10](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88Lm-K7wzWR7rXT_ss:-L88Lm-K7wzWR7rXT_st:b4lv4gf) ([raw file](https://github.com/batfish/batfish/blob/57dcf3603e4ba241a90a09658320ca493eb8bea7/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestIpSpace.java#L6)):*
> ```Java
> import org.batfish.datamodel.visitors.GenericIpSpaceVisitor;
> 
> public class TestIpSpace implements IpSpace {
> ```

MockIpSpace?

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestRib.java, line 10 at r10](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88Lo937mBEBrR35uPL:-L88Lo937mBEBrR35uPM:bqmog2r) ([raw file](https://github.com/batfish/batfish/blob/57dcf3603e4ba241a90a09658320ca493eb8bea7/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestRib.java#L10)):*
> ```Java
> import java.util.SortedSet;
> 
> public class TestRib implements GenericRib<AbstractRoute> {
> ```

MockRib?

---

*[projects/batfish/src/test/java/org/batfish/z3/TestSynthesizerInput.java, line 16 at r5](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88MAsH4c2MrItRflkP:-L88MAsH4c2MrItRflkQ:b-2khcsy) ([raw file](https://github.com/batfish/batfish/blob/1a569e10fea2025f44ee8394bbc40b4616961768/projects/batfish/src/test/java/org/batfish/z3/TestSynthesizerInput.java#L16)):*
> ```Java
> import org.batfish.z3.state.StateParameter.Type;
> 
> public class TestSynthesizerInput implements SynthesizerInput {
> ```

MockSynthesizerInput?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/AclIpSpace.java, line 85 at r10](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88K_PbCx5TnozCUcRD:-L88K_PbCx5TnozCUcRE:b-bndk5d) ([raw file](https://github.com/batfish/batfish/blob/57dcf3603e4ba241a90a09658320ca493eb8bea7/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/AclIpSpace.java#L85)):*
> ```Java
>   @Override
>   public int hashCode() {
>     return Objects.hash(_lines);
> ```

cache the hash?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 57 of 78 files reviewed at latest revision, 9 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java, line 31 at r10](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88LEwSCiEoMRrT9CZM:-L88o--F4_CcpUvfXJpk:bvfq8zq) ([raw file](https://github.com/batfish/batfish/blob/57dcf3603e4ba241a90a09658320ca493eb8bea7/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java#L31)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

what is this?
</blockquote></details>

The primary address plus secondary addresses. Interfaces are not restricted to a single IP nor a single subnet.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestArpAnalysis.java, line 7 at r6](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88LgPtCO-b97GAACPg:-L88pwhg7znKQg0YwB5g:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/3e00658c6667934835c8270e6f5979b67f2d0d2d/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestArpAnalysis.java#L7)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

I expected to see tests here. Maybe rename it `MockArpAnalysis`?
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestIpSpace.java, line 6 at r10](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88Lm-K7wzWR7rXT_ss:-L88pz6Q0imhysCJCCbi:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/57dcf3603e4ba241a90a09658320ca493eb8bea7/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestIpSpace.java#L6)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

MockIpSpace?
</blockquote></details>

Done.

---

*[projects/batfish/src/test/java/org/batfish/z3/TestSynthesizerInput.java, line 16 at r5](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88MAsH4c2MrItRflkP:-L88q0Of5GVHjy2FqVPy:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/1a569e10fea2025f44ee8394bbc40b4616961768/projects/batfish/src/test/java/org/batfish/z3/TestSynthesizerInput.java#L16)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

MockSynthesizerInput?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 56 of 78 files reviewed at latest revision, 9 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/AclIpSpace.java, line 85 at r10](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88K_PbCx5TnozCUcRD:-L88qk1h-LESv8gyc4Ka:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/57dcf3603e4ba241a90a09658320ca493eb8bea7/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/AclIpSpace.java#L85)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

cache the hash?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 56 of 78 files reviewed at latest revision, 9 unresolved discussions.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestRib.java, line 10 at r10](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88Lo937mBEBrR35uPL:-L88qyU58T2KJUBTmZW_:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/57dcf3603e4ba241a90a09658320ca493eb8bea7/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/TestRib.java#L10)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

MockRib?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 56 of 78 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[a discussion](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L88rgYB-lS6FZFQP3PK:-L88rgYB-lS6FZFQP3PL:b-gnird7) (no related file):*
TODO: choose better name for `ArpAnalysis` and related classes.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 1 of 34 files at r10, 26 of 41 files at r11, 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java, line 54 at r12](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L89GTnMDWrCoHCisweC:-L89GTnMDWrCoHCisweD:bq8afnp) ([raw file](https://github.com/batfish/batfish/blob/413732db19ecc197181ded219125bb0eab9ef212/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java#L54)):*
> ```Java
>     private Configuration _owner;
> 
>     private Set<InterfaceAddress> _secondaryAddresses;
> ```

comment?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 16 of 33 files at r13.
Review status: 77 of 91 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 14 of 33 files at r13.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 2 of 2 files at r14, 8 of 8 files at r15.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java, line 213 at r15](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L8DoluFEOf3b_ui75pY:-L8DoluFEOf3b_ui75pZ:b-72egtv) ([raw file](https://github.com/batfish/batfish/blob/ef38a892469d9613bbc9b226290743780d9d1f80/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java#L213)):*
> ```Java
>      * set containing the primary address and secondary addresses.
>      */
>     public Builder setSecondaryAddresses(Iterable<InterfaceAddress> secondaryAddresses) {
> ```

But what are secondary addresses?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 3 of 6 files at r16.
Review status: 94 of 97 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 3 of 6 files at r16.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java, line 213 at r15](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L8DoluFEOf3b_ui75pY:-L8DyfR7BkFf60brgPwd:b-rqo697) ([raw file](https://github.com/batfish/batfish/blob/ef38a892469d9613bbc9b226290743780d9d1f80/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java#L213)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

But what are secondary addresses?
</blockquote></details>

A secondary address is another address potentially associated with a different subnet living on the interface. The primary address is the one used by default as the source IP for traffic sent out the interface. The node will accept traffic whose destination IP belongs is among any of the addresses of any of the interfaces.
The interface will reply to ARP for the primary or any secondary IP.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java, line 213 at r15](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L8DoluFEOf3b_ui75pY:-L8E-0_o-z5-OfxNqmfP:b-9129am) ([raw file](https://github.com/batfish/batfish/blob/ef38a892469d9613bbc9b226290743780d9d1f80/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java#L213)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

A secondary address is another address potentially associated with a different subnet living on the interface. The primary address is the one used by default as the source IP for traffic sent out the interface. The node will accept traffic whose destination IP belongs is among any of the addresses of any of the interfaces.
The interface will reply to ARP for the primary or any secondary IP.
</blockquote></details>

put that in the comment please

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 94 of 101 files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java, line 213 at r15](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L8DoluFEOf3b_ui75pY:-L8E69agESTr93sQsXWR:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/ef38a892469d9613bbc9b226290743780d9d1f80/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Interface.java#L213)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

put that in the comment please
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 7 of 7 files at r17.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 1 of 1 files at r18.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java, line 306 at r18](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L8EjYjp-g0KVGDKzJV4:-L8EjYjp-g0KVGDKzJV5:bq46flc) ([raw file](https://github.com/batfish/batfish/blob/15480bae503006b762a278c469885110bdc52ab6/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java#L306)):*
> ```Java
>         });
> 
>     return output
> ```

this is just a deep copy?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 100 of 105 files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java, line 306 at r18](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L8EjYjp-g0KVGDKzJV4:-L8Ek58-4HhmMEw0RYBV:b3goskr) ([raw file](https://github.com/batfish/batfish/blob/15480bae503006b762a278c469885110bdc52ab6/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java#L306)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

this is just a deep copy?
</blockquote></details>

Yes, but into more compatct Immutable structures

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 15 of 21 files at r19.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 102 of 104 files reviewed at latest revision, 1 unresolved discussion.

---

*[projects/checkstyle.xml, line 43 at r9](https://beta.reviewable.io/reviews/batfish/batfish/1047#-L83O4xWvioyHSLOapnW:-L8FLxlhD04-PP3cqBR8:btr69ej) ([raw file](https://github.com/batfish/batfish/blob/2e02a47764654cbe24fdf14346f0ce73ac7c967a/projects/checkstyle.xml#L43)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Tomato potato
</blockquote></details>

Fixed by you PR

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
When it looks good to Matt, looks good to me -- no need for another reviewer, as it's already had one detailed reviewer and user. The rest will come out in the wash.

---

Review status: 102 of 104 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1047#-:-L8GNKwG69xSXO-no56t:b-fhzybt)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 3 of 3 files at r20.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1047)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1047#-:-L8Ivn_yokz0jWZmOQ49:bnfp4nl)*
<!-- Sent from Reviewable.io -->
